sdk/core/azure-core/src/main/java/com/azure/core/util/QueryParameter.java

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Represents a single parameter to be added to a query string.
+ *<p>
+ * If multiple values are added to a query string with the same name (case-insensitive), then the values will be
+ * appended at the end of the same {@link QueryParameter} with commas separating them.
+ */
+public final class QueryParameter {
```

Comment:
At this point I'm not sure there is any need for this to be public API any longer? I think it is only used as an implementation detail and could therefore be made package-private or moved into an implementation package?

Reply:
Hi @JonathanGiles,

`getQuery()` provides a high-level view of what is going to be in the query string, so we can have more code like this:

``` 
// from Storage SDK
if (urlBuilder.getQuery().containsKey(SIG)) {
   urlBuilder.setQueryParameter(SIG, "REDACTED");
```

and less like this:

```
// also from Storage SDK
 request.url.getQuery().contains("comp=block") &&
                !request.url.getQuery().contains("comp=blocklist")
```

Since we are adding multiple parameter values (of the same name) to `UrlBuilder` we'd have to break `getQuery()` and make it return `Map<String, List<String>>` instead. The suggestion then by @alzimmermsft was to maintain `getQuery()` and add a `QueryParameter` class and correspondent `getQueryParameters()` to provide a higher-level view (from `String`) into the state of the `UrlBuilder` when it comes to the query string. The idea was also to use `QueryParameter` to provide more services in the future, in the same way that `Header` is used. 

Right now, it does not have clients, granted, but it may be a good idea to keep `getQueryParameters()` public if 1) we want clients to start using it to have high-level access to what's going to be in the query string, correct with multiple values and all 2) if we need it in AutoRest (which is why this modification is being made). In short, if high-level access is needed to the key-pair values that will be in the query string, then `QueryParameter` is a good way to do it, because it's an implementation doing the right thing with multiple query parameters - and also, if we want to add new services to it later, we wouldn't need to submit another PR just to make it public.

If that's not enough reason, I can definitely move it to an implementation package and make it public with another PR later if needed. 

Please let me know @JonathanGiles @alzimmermsft @srnagar so we can close and merge this one. Thanks!

Reply:
I might be misunderstanding something - you speak of `getQueryParameters()` as if it exists, but I don't see it?

Reply:
@JonathanGiles `getQueryParameters()` was removed in one of yesterday's commits, as no packages were using it at this time. With that removal, then yes, `QueryParameter` should probably be made package visible only. If we want to expose `getQueryParameters()` for a higher-level view of the query string parameters (higher than `getQuery()` which is String, String) then we'd have to keep `QueryParameter` public.

So it's a package deal as I see it: public `QueryParameter` plus `getQueryParameter()` or making `QueryParameter` private and leaving `getQueryParameters` out. 

Your observation on `QueryParameter` earlier was right: it's right now only used privately. Leaving it in would be more about possible uses in the future. 

Please let me know your thoughts on which way to take here. Cheers!

Reply:
Relevant commit: https://github.com/Azure/azure-sdk-for-java/pull/21203/commits/12eaf87f547aeb8ba1b311e59768990f26790893

Reply:
Got it, thanks for clarifying. I don't have a strong sense for how often someone is going to want to enumerate each of the values for a single query parameter, but I guess it would be a non-zero amount of the time. We could eagerly add this API or hold back until we think we have things right. By default I would hold back on any new API until someone is really asking for it (which could be the end user or the person developing the API if they have a strong sense). If you think that there is value in having this from the get-go, let's make it available, but otherwise lets make this package-private / implementation for now.

Reply:
Sounds good to me. I will make the change to `QueryParameter` then and ping back here once it's gone through the checks. Cheers!